### PR TITLE
fix: webui update metrics to opt-out by default (part of 2074)

### PR DESCRIPTION
**Pull request by djmcquillan**
fixes: https://github.com/ipfs/ipfs-desktop/issues/2370
fixes: #2085 

- Updates  banner per discussion ([here](https://www.notion.so/pl-strflt/Metrics-Collection-and-Consent-Language-0d4059f11d4d474bb76d00539d778d8d#35482f9eef47464cb9efab8eaf7b4f8f)) and enables metrics on app load

### CODE REVIEW

File: public/locales/en/status.json
Comment (SgtPooki) [inline public/locales/en/status.json:17]: @juliaxbow for visibility
cc @tinytb @lidel

File: src/bundles/analytics.js
Comment (SgtPooki) [inline src/bundles/analytics.js:166]: this comment is invalid

File: src/bundles/analytics.js
Comment (SgtPooki) [inline src/bundles/analytics.js:322]: I would change this comment to `add consent by default`

File: src/status/StatusPage.js
Comment (SgtPooki) [inline src/status/StatusPage.js:60]: I just want to confirm: metrics will have already been enabled prior to user clicking anything on the banner, yes?

File: src/status/StatusPage.js
Comment (SgtPooki) [inline src/status/StatusPage.js:61]: we could probably remove the strings for these unused items as well

**Review by SgtPooki — CHANGES_REQUESTED** [review]

File: public/locales/en/status.json
Comment (djmcquillan) [reply public/locales/en/status.json:17]: as another suggestion here we could follow the same verbiage being used on public gateway checker:
"This site collects metrics and other telemetry data in accordance with our metrics collection policy. You can learn more about how we use this information and customize your preferences by clicking “manage settings” (link to settings page).

File: src/bundles/analytics.js
Comment (SgtPooki) [inline src/bundles/analytics.js:413]:
```suggestion
        case ACTIONS.SET_ANALYTICS_DISPLAY_BANNER: {
          const showAnalyticsBanner = action.payload?.shouldShow || false
          return { ...state, showAnalyticsBanner }
```

File: src/bundles/analytics.js
Comment (SgtPooki) [inline src/bundles/analytics.js:292]:
```suggestion
      ACTIONS.SET_DISPLAY_ANALYTICS_DISPLAY_BANNER
```
I think this name would make better sense as "SET_DISPLAY_ANALYTICS_DISPLAY_BANNER," what do you think? we're not technically asking users to enable metrics, because we're enabling them by default.

File: src/bundles/analytics.js
Comment (SgtPooki) [inline src/bundles/analytics.js:81]:
```suggestion
  'SET_DISPLAY_ANALYTICS_DISPLAY_BANNER'
```

File: src/bundles/analytics.js
Comment (SgtPooki) [inline src/bundles/analytics.js:53]:
```suggestion
 * @property {'SET_DISPLAY_ANALYTICS_DISPLAY_BANNER'} type
```
can we change this to be more relevant to what we're doing with the banner now? maybe `SET_DISPLAY_ANALYTICS_DISPLAY_BANNER` ?

**Review by SgtPooki — CHANGES_REQUESTED** [review]
some naming confusion, and targeting of specific usecases with the tests.

File: src/bundles/analytics.test.js
Comment (SgtPooki) [inline src/bundles/analytics.test.js:35]: we should explicitly test each of these. If store has X val, analyticsEnabled should be true, initialConsent should be true.. if store has Y val, one true one false.. etc.

File: src/bundles/analytics.js
Comment (djmcquillan) [reply src/bundles/analytics.js]: we can use lastDisabledAt > lastEnabledAt as a check for past users prior to these new updates.  from this patch on we can rely on the new flag optedOut and consent.length

File: src/bundles/analytics.test.js
Comment (djmcquillan) [reply src/bundles/analytics.test.js]: tests have been cleaned up

File: src/bundles/analytics.test.js
Comment (djmcquillan) [reply src/bundles/analytics.test.js]: agreed I have added new tests to better match the refactor

File: src/bundles/analytics.js
Comment (djmcquillan) [reply src/bundles/analytics.js]: updates to clearer names to match the refactor have been pushed

File: src/bundles/analytics.js
Comment (SgtPooki) [inline src/bundles/analytics.js:411]:
```suggestion
          return { ...state, lastDisabledAt: 0, lastEnabledAt: 0, consent, optedOut: didOptOutCompletely, showAnalyticsBanner: false }
```

File: src/bundles/analytics.test.js
Comment (SgtPooki) [inline src/bundles/analytics.test.js:33]: :+1:

File: src/bundles/analytics.test.js
Comment (SgtPooki) [inline src/bundles/analytics.test.js:36]: remove

File: src/bundles/analytics.test.js
Comment (SgtPooki) [inline src/bundles/analytics.test.js:44]: :+1:

File: src/bundles/analytics.test.js
Comment (SgtPooki) [inline src/bundles/analytics.test.js:58]: :+1:

File: src/bundles/analytics.test.js
Comment (SgtPooki) [inline src/bundles/analytics.test.js:73]: nit: technically, the name here should be "...prior to new opt-out requirement" because the policy of telemetry is "opt-out" which means users are implicitly granting permissions. The action they have to perform to change the default is "opt-out". 

The name test is :+1: though.

File: src/bundles/analytics.test.js
Comment (SgtPooki) [inline src/bundles/analytics.test.js:88]: :+1:

File: src/bundles/analytics.test.js
Comment (SgtPooki) [inline src/bundles/analytics.test.js:62]: we should add a test here where lastEnabledAt also has a date.

File: src/bundles/analytics.test.js
Comment (SgtPooki) [inline src/bundles/analytics.test.js:107]: :+1:

File: src/bundles/analytics.test.js
Comment (SgtPooki) [inline src/bundles/analytics.test.js:116]: :+1:

File: src/bundles/analytics.test.js
Comment (SgtPooki) [inline src/bundles/analytics.test.js:138]: make it easy to update, and see what the actual dates are.
```suggestion
      lastEnabledAt: (new Date('2022-01-01')).getTime(),
      lastDisabledAt: (new Date('2022-01-02')).getTime(),
```

File: src/bundles/analytics.test.js
Comment (SgtPooki) [inline src/bundles/analytics.test.js:134]: :+1:

File: src/bundles/analytics.test.js
Comment (SgtPooki) [inline src/bundles/analytics.test.js:130]:
```suggestion
    expect(store.selectAnalyticsConsent()).toEqual([])
    expect(store.selectAnalyticsOptedOut()).toBe(true)
```

File: src/bundles/analytics.test.js
Comment (SgtPooki) [inline src/bundles/analytics.test.js:135]: :+1:

File: src/bundles/analytics.test.js
Comment (SgtPooki) [inline src/bundles/analytics.test.js:158]: use explicit dates for tests to make things easier for us in the future.

```suggestion
      lastEnabledAt: (new Date('2022-01-02')).getTime(),
      lastDisabledAt: (new Date('2022-01-01')).getTime(), // past
```

File: src/bundles/analytics.test.js
Comment (SgtPooki) [inline src/bundles/analytics.test.js:170]:
```suggestion
      lastEnabledAt: (new Date('2022-01-02')).getTime(),
      lastDisabledAt: (new Date('2022-01-01')).getTime(), // or zero
```

File: src/components/analytics-banner/AnalyticsBanner.js
Comment (SgtPooki) [inline src/components/analytics-banner/AnalyticsBanner.js:5]: I thought this was a new file but it looks like it's just a copy of `askToEnable`. Awesome.. thanks for the rename

File: src/bundles/analytics.js
Comment (SgtPooki) [reply src/bundles/analytics.js:161]: After looking through the rest of the code.. I'm sure that things will work with this test, but concerned that we're not checking for either

1. `!state.analytics.optedOut`
2. or setting lastDisabledAt & lastEnabledAt to 0.

With the current code.. the only implicit blocking of this returning true again in the future is that the call to `selectAnalyticsOptedOutPriorToDefaultOptIn` is blocked by a call to `selectAnalyticsOptedOut`. However, if someone is opted out, then enables consent, then disables consent, then we will show the banner again, won't we? Is there already a test for that?

Comment (SgtPooki) [reply src/bundles/analytics.js:161]: i think we just need to have a check for optedout. `if !state.analytics.optedOut && consent === 0 && lastDisabledAt > lastEnabledAt`

| State name | lastDisabledAt | lastEnabledAt | optedOut | !optedOut && lastDisabledAt > lastEnabledAt | show banner? |
|------------|:--------------:|---------------|----------|---------------------------------------------|--------------|
| Disabled   | 2022-01-01     | 0             | false    | true                                        | true         |
| Enabled    | 2022-01-01     | 2022-01-02    | false    | false                                       | false        |
| Disabled   | 2022-01-03     | 2022-01-02    | true     | false                                       | false        |

Comment (djmcquillan) [reply src/bundles/analytics.js:161]: thank you for the detailed charts, I do believe all of these scenarios are covered because we have a check in the init that state.analytics.optedOut and within that condition if they have opted out we essentially do nothing, if they have not opted out and previously disabled (row 1 in table above) then that is the only time the showAnalyticsBanner is set to true otherwise the showAnalyticsBanner is reset to false during other actions (should maybe be moved or more explicit in the code)

I am pretty sure tests will cover all of the above scenarios but there are a lot of them! basically I used lastDisabledAt + lastEnabledAt + consent.length for all existing and previous logic, then added the two new flags optedOut and showAnalyticsBanner to cover future scenarios of users toggling on/off consent

File: src/bundles/analytics.js
Comment (djmcquillan) [reply src/bundles/analytics.js:402]: it should be set by action creator dispatching ANALYTICS_DISABLED
`  doDisableAnalytics: () => ({ dispatch, store }) => {
    root.Countly.opt_out()
    removeConsent(consentGroups.all, store)
    dispatch({ type: 'ANALYTICS_DISABLED', payload: { consent: [] } })
  },`

File: src/bundles/analytics.js
Comment (djmcquillan) [reply src/bundles/analytics.js:400]: I think we could but I did not want to mess too much with the existing logic as it seemed to also be used in the default of the reducer

File: src/bundles/analytics.js
Comment (SgtPooki) [inline src/bundles/analytics.js:161]: this was the only change to code I was expecting. we should revert the changes to lastEnabledAt and lastDisabledAt

File: src/bundles/analytics.js
Comment (SgtPooki) [inline src/bundles/analytics.js]: didn't we discuss not modifying these and only adding the optedOut check to `selectAnalyticsOptedOutPriorToDefaultOptIn`

**Review by SgtPooki — CHANGES_REQUESTED** [review]

File: src/bundles/analytics.js
Comment (djmcquillan) [reply src/bundles/analytics.js:405]: after discussing offline we will leave lastEnabledAt and lastDisabledAt as is for now to keep existing logic working as it previously was

File: src/bundles/analytics.test.js
Comment (djmcquillan) [reply src/bundles/analytics.test.js:130]: added as part of latest commit

File: src/bundles/analytics.js
Comment (djmcquillan) [reply src/bundles/analytics.js:400]: just commented here based on offline conversation we will leave lastEnabledAt and lastDisabledAt as is

File: src/bundles/analytics.js
Comment (djmcquillan) [reply src/bundles/analytics.js]: yea reverted and added some comments

**Review by SgtPooki — APPROVED** [review]
I'm good to merge this if we get approval from Nishant or Lidel as well. A new webui release should wait on new privacy policy though.

**Comment by SgtPooki** [comment]
> @SgtPooki we should host PRs on fleek, like public-gateway-checker does. We already host this on https://webui.ipfs.io/ this way we can test PR specific changes too.

yes we should

**Pull request merged**
